lib/collection/src/collection/query.rs

```diff
@@ -0,0 +1,122 @@
+use std::sync::Arc;
+use std::{iter, mem};
+
+use futures::{future, TryFutureExt};
+use itertools::{Either, Itertools};
+use segment::types::Order;
+use segment::utils::scored_point_ties::ScoredPointTies;
+
+use super::Collection;
+use crate::operations::consistency_params::ReadConsistency;
+use crate::operations::shard_selector_internal::ShardSelectorInternal;
+use crate::operations::types::CollectionResult;
+use crate::operations::universal_query::shard_query::{
+    Fusion, ScoringQuery, ShardQueryRequest, ShardQueryResponse,
+};
+
+struct IntermediateQueryInfo<'a> {
+    scoring_query: Option<&'a ScoringQuery>,
+    take: usize,
+}
+
+impl Collection {
+    async fn query_shards_concurrently(
+        &self,
+        request: Arc<ShardQueryRequest>,
+        read_consistency: Option<ReadConsistency>,
+        shard_selection: &ShardSelectorInternal,
+    ) -> CollectionResult<Vec<ShardQueryResponse>> {
+        // query all shards concurrently
+        let shard_holder = self.shards_holder.read().await;
+        let target_shards = shard_holder.select_shards(shard_selection)?;
+        let all_searches = target_shards.iter().map(|(shard, shard_key)| {
+            let shard_key = shard_key.cloned();
+            shard
+                .query(
+                    Arc::clone(&request),
+                    read_consistency,
+                    shard_selection.is_shard_id(),
+                )
+                .and_then(move |mut records| async move {
+                    if shard_key.is_none() {
+                        return Ok(records);
+                    }
+                    for batch in &mut records {
+                        for point in batch {
+                            point.shard_key.clone_from(&shard_key);
+                        }
+                    }
+                    Ok(records)
+                })
+        });
+        future::try_join_all(all_searches).await
+    }
+
+    /// This is a remote shard implementation to querying the shards. Only used for the internal service
+    pub async fn query_internal(
```

Comment:
tbh, I am struggling to understand what's going on in this function. it might require some more comments/decomposition

Reply:
This is what will be called on the remote shard. So that, if it is a fusion request, it sends the merged prefetches instead of the main query. 

I'll make it easier to look at